# JBrowse 2 visualizations gallery

Below are some examples of the visualizations JBrowse 2 has to offer. For live
demos, have a look at the [demos page](../demos).

---

![](/img/dotplot.png)

A dotplot showing the alignment between the grape vs peach genome, loaded from a
PAF file generated by minimap2
([live link](https://jbrowse.org/code/jb2/latest/?config=test_data%2Fconfig_dotplot.json&session=share-r4sMB3bHh5&password=C9jCa))

---

![](/img/linear_synteny.png)

A linear synteny view of the grape vs peach genome using data from
[MCScan](<https://github.com/tanghaibao/jcvi/wiki/MCscan-(Python-version)>). The
screenshot shows both the individual connections between aligned protein coding
genes, and the larger synteny blocks. The synteny blocks are also rendered as
the blue (inverted) and red (non-inverted) blocks on each linear genome view
([live link](https://jbrowse.org/code/jb2/latest/?config=test_data%2Fconfig_dotplot.json&session=share-4MjF5YGM_G&password=rByjt))

---

![](/img/hic_track.png)

Screenshot showing a Hi-C contact matrix rendered in JBrowse using the .hic file
format (from Juicebox)
([live link](http://jbrowse.org/code/jb2/latest/?config=test_data%2Fconfig_demo.json&session=share-xS8Eg67AFS&password=jPzH5))

---

![](/img/sv_inspector_importform_loaded.png)

Screenshot of the SV inspector displaying inter-chromosomal translocations in
the SKBR3 cell line. The SV inspector is a tabular view plus whole-genome
circular overview of the SVs
([live link](https://jbrowse.org/code/jb2/latest/?config=test_data%2Fconfig_demo.json&session=share-n9_vE%2FEl2R&password=wu9J6))

---

![](/img/horizontally_flip.png)

Screenshot showing the ability to "horizontally flip" the linear genome view.
This feature allows the genome to be read "right to left", or 3' to 5' instead
of 5' to 3', which can help when looking at anti-sense genes or features
([live link](http://jbrowse.org/code/jb2/latest/?config=test_data%2Fconfig_demo.json&session=share-6pkcSXlbFL&password=ER28C))

---

![](/img/cnv.png)

This screenshot shows whole-genome sequencing coverage from the COLO829 melanoma
cancer cell line. It shows both "normal" (blue) and "tumor" (red) sequencing
coverage on the same track (a
["multi-quantitative"](/docs/user_guides/multiquantitative_track/) track). The
sequencing coverage is computed from mosdepth and stored in BigWig files
([live link](https://jbrowse.org/code/jb2/latest/?config=test_data%2Fconfig_demo.json&session=share-AcZSrC_yOb&password=e7b64))

---

![](/img/skbr3_translocation.png)

Image of a possible translocation in the SKBR3 dataset using the "breakpoint
split view" which can connect supporting reads (black curves) and the variant
call itself (green thicker line, with feet indicating directionality)
([live link](https://jbrowse.org/code/jb2/latest/?config=test_data%2Fconfig_demo.json&session=share-Swq8pJTX0z&password=yM41l))

---

![](/img/smalldel.png)

Screenshot showing a heterozygous small deletion in the GIAB dataset. The
nanopore reads are colored and sorted by the HP (haplotype) tag which helps show
that the deletion is exclusive to one haplotype. Users can color, sort, group,
and filter by tags
([live link](https://jbrowse.org/code/jb2/latest/?config=test_data%2Fconfig_demo.json&session=share-psOr2x2efp&password=bErZE))

---

![](/img/insertion.png)

Screenshot showing a ~1.5kb insertion in the GIAB dataset, with Nanopore,
PacBio, and Illumina reads. This screenshot shows the "Show soft clipping"
option enabled on the Illumina reads, which helps reveal the boundaries of the
insertion on the short reads, while the long reads fully span the insertion
([live link](https://jbrowse.org/code/jb2/latest/?config=test_data/config_demo.json&session=share-oTyYRpz9fN&password=fYAbt))

---

![](/img/read_vs_ref_insertion.png)

Screenshot showing a ~500bp insertion from SKBR3 (breast cancer) cell line
PacBio reads. Users can right-click sequence reads, then click "read vs
reference" to show complex read alignments in a "synteny style" view (lower
panel). Users can also interactively get the sequence of read portions with a
click-and-drag action (lower panel, in blue)
([live link](https://jbrowse.org/code/jb2/latest/?config=test_data%2Fconfig_demo.json&session=share-rzJ27iixQH&password=rSgZe))

---

![](/img/modifications.png)

A methylated and unmethylated CpG island shown using the MM tag with nanopore
reads on the COLO829 (melanoma) cancer cell line.
([live link](https://jbrowse.org/code/jb2/latest/?config=test_data%2Fconfig_demo.json&session=share-LffYr8SI5E&password=VmZVl))

---

![](/img/multisv.png)

An example of the 1000 genomes structural variant call VCF, showing a large
inversion on chromosome 19. This screenshot used the in-app clustering workflow
on the inverted region, and then zoomed out further
([live link](https://jbrowse.org/code/jb2/latest/?config=%2Fgenomes%2FGRCh38%2F1000genomes%2Fconfig_1000genomes.json&session=share-DN_h4SIwo4&password=CxkLw))
