We are excited to announce the latest JBrowse release! This has many great
features including

1. Floating feature labels, so if you are zoomed into the middle of a gene, the
   feature label hangs out alongside it on the left side of the screen! The
   results may not be pixel perfect in all cases (reverse complement labels can
   end up overallping other features for example) but we hope this helps you
   see the labels for features that extend off the screen!

   ![](https://user-images.githubusercontent.com/6511937/163470981-cfbd4464-bd5a-4421-8d9c-c8e6bb2d19bc.png)

   Figure showing the feature labels floating alongside gene even when you are
   zoomed into the feature

2. Upgraded build system for jbrowse-web and jbrowse-desktop to webpack 5, and
   added examples of using webpack 5 with embedded components. Webpack 5
   results in smaller build sizes for jbrowse-web. For a typical session,
   amount of .js downloaded with gzip enabled in v1.6.9 is ~1.4MB, in v1.7.0
   ~900kb (-500kb smaller). Without gzip enabled v1.6.9 4.8MB, v1.7.0 2.5MB
   (-2MB smaller). The result is largely due to better webworker bundling.

3. Optimizations and improved visualization of paired-end reads in the
   breakpoint split view

4. Optimized linear synteny renderings with long CIGAR strings. These
   optimizations help viewing large alignments such as the CHM13-T2T-v2.0
   assembly vs hg38 PAF file from https://github.com/marbl/CHM13#downloads
   (example here
   https://jbrowse.org/code/jb2/main/?config=test_data%2Fconfig_demo.json&session=share-0qD2-d_k4K&password=GF8Sk)

5. The "Assembly selector" dropdown box now remembers your last selection, so
   you can more easily return to your genome of interest
