Hello all!

This release announces v3.0.0 of JBrowse 2. It is not at all a dramatic change
like the JBrowse 1 -> JBrowse 2 transition, it's much more incremental, and in
fact, you can go on calling it JBrowse 2...as a point of comparison, ggplot2 is
on v3.5.1 right now :)

We decided to make a major bump due to some small "breaking" changes that could
affect plugin and embedded library users in particular.

But for most users, this release should have nothing but the usual bugfixes and
improvements! And there are a lot of these!

Without further ado, here are some highlights!

## Improved rendering of phased VCF in multi-variant view

Phased variants offer a unique opportunity to see which parent particular
variants came from. To help this, we added a new "phased" rendering mode to the
multi-variant renderings, that takes phased VCF and renders each phase as an
individual row. This allows users to effectively see the entire row as a
haplotype, and can be used to help visualize trio VCF to find parent of origin

![Screenshot From 2025-01-25 15-03-21](https://github.com/user-attachments/assets/a9308a40-ab74-48c0-9ab6-035b50a1ae0b)
Screenshot showing the "phased mode" for a trio VCF

![image](https://github.com/user-attachments/assets/4198d6ba-05fd-4be7-829e-76f58499ad46)

Screenshot showing how to access the phased rendering mode

## Rendring of phase-sets

This could be more or less uncommon depending on your work, but for incompletely
phased variant calls you will see "phased sets" (PS tag in genotypes). We color
each set a random color

![image](https://github.com/user-attachments/assets/63fa9f59-d54b-4f66-852d-4d8592ff95ec)

Screenshot showing phase set rendering of a phased VCF. This is nearly
completely phased so just showing green, but each phase set would get a unique
color. The black entries are unphased variants (which could be filtered out, but
this is not done currently)

## Rendering polyploids in the multi-variant view

![image](https://github.com/user-attachments/assets/58138b02-4d20-487b-af77-68b2d038dedc)

Screenshot showing the "polyploid" rendering of the multi-variant display.

Polyploid variant calls can look like `0/2/./1` indicating one match to the
reference (0), two alts (1,2), and one missing call (.).

- Yellow indicates missingness
- Grey indicates reference
- Blue indicates the ALT===1
- Red is ALT!==1

Each of these is a spectrum that increases in darkness, so darker blue indicates
more ALT===1, darker red indicates more ALT!==1, darker yellow is more uncalled,
etc.

## Adding sample metadata to VCF from an external TSV

![image](https://github.com/user-attachments/assets/ba7b99ab-53eb-4719-a6de-1cf9ff16511a)

Screnshot showing multi-sample matrix display, with the population colors coming
from a "sample TSV" file. The below config shows the samplesTsvLocation config,
the first column should match the sample name in the VCF

Example TSV

```
name	population
HG01879	ACB
HG01880	ACB
HG01881	ACB
HG01882	ACB
HG01883	ACB
HG01884	ACB
```

Config

```json
{
  "type": "VariantTrack",
  "trackId": "1kGP_high_coverage_Illumina.chr1.filtered.SNV_INDEL_SV_phased_panel.vcf",
  "name": "1kGP_high_coverage_Illumina.chr1.filtered.SNV_INDEL_SV_phased_panel.vcf",
  "assemblyNames": ["hg19"],
  "adapter": {
    "type": "VcfTabixAdapter",
    "vcfGzLocation": {
      "uri": "https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/release/20130502/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"
    },
    "index": {
      "location": {
        "uri": "https://ftp-trace.ncbi.nlm.nih.gov/1000genomes/ftp/release/20130502/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi"
      }
    },
    "samplesTsvLocation": {
      "uri": "https://jbrowse.org/genomes/hg19/1000g.sorted.csv.gz"
    }
  }
}
```

## Improved UI for opening synteny tracks

It has always been challenging for end users to open synteny tracks

To help with this, we now support adding synteny tracks via the default "Add
track" workflow, and make it easier to specify the query and target assembly
names. We also added a similar UI change to the "Linear synteny view" import
form

![image](https://github.com/user-attachments/assets/913fd8b9-9d4e-4d44-88d7-ba3e70fd54f4)

Screnshot showing the default add track workflow with new UI for choosing target
and query assemblies for PAF tracks

![image](https://github.com/user-attachments/assets/75da5058-50cc-413e-aee9-e57362d2d0c0)

Screenshot showing the add track workflow in the "Linear synteny view" import
form, also showing new UI for choosing query and target assmblies

## New "turbo zoom" buttons

In working with users, we noticed people having to click the zoom buttons or
fiddle with the slider a lot. To aid this, we added a small dropdown for quickly
doing 10x, 50x, and 100x zooms

![image](https://github.com/user-attachments/assets/2ee0305e-d617-4ec3-8957-8ec1d906feff)

## For developers: some "breaking" changes

1. We changed the "filehandle" type used in data adapters. We now use
   generic-filehandle2 which is simpler and does not require a node.js polyfill
   on the web
2. Upgrading React 18 -> React 19 in our main webapp. This drops support for
   React 17, and may have some other implications

If you are a plugin or embedded user, and experience problems with upgrading,
let us know!
