This release adds several interesting features including

## Sticky view headers and pinned tracks

We created a new UI where the "view header" stays visible even when scrolling
the page.

We also added the ability to "pin" tracks in the linear genome view, which also
stay visible when scrolling the page.

## Cluster multi-quantitative tracks

Motivated by the clustering workflow in the Multi-variant viewer, a similar
feature was added to the multi-quantitative track

To do this, we sample the data according to the current zoom level (otherwise,
it would create very large matrices at per-base resolution) and then provide an
R script the user can run to cluster the data

Before clustering
![image](https://github.com/user-attachments/assets/a8c64fb8-7bb4-44a1-8ea0-b54a0d88e988)

After clustering
![Screenshot From 2025-03-11 00-25-06](https://github.com/user-attachments/assets/ff026b8a-e46e-4f66-afd1-314841b4d586)

## Simplify some adapter configuration

The configuration system has a lot of complex substructure, but in this release
we created the ability to supply a simplified config, and the results will be
auto-determined

For example

```json
{
  "type": "AlignmentsTrack",
  "trackId": "volvox_alignments",
  "name": "volvox-sorted.bam",
  "assemblyNames": ["volvox"],
  "adapter": {
    "type": "BedGraphTabixAdapter",
    "bedGraphGzLocation": {
      "uri": "file.bed.gz"
    },
    "index": {
      "location": {
        "uri": "file.bed.gz.tbi"
      }
    }
  }
}
```

after

```json
{
  "type": "AlignmentsTrack",
  "trackId": "volvox_alignments",
  "name": "volvox-sorted.bam",
  "assemblyNames": ["volvox"],
  "adapter": {
    "type": "BedGraphTabixAdapter",
    "uri": "file.bed.gz"
  }
}
```

The CLI tools and other systems will continue to output the verbose version for
the time being, but this simplified system will be handy to anyone who does
hand-coding of the configuration
