This is a big release on a number of fronts!

## JBrowse 2 genome hubs

We created a new website called https://genomes.jbrowse.org that has new JBrowse
2 instances for over 50,000 species. This massive effort is primarily driven by
bulk loading data from UCSC, both their main browsers and GenArk hubs which has
data for thousands of NCBI RefSeq (curated) and GenBank genomes

All of these JBrowse 2 instances are available on both the web and on the
desktop app

Users can even browse "synteny views" between the different hubs by opening a
liftOver track, see https://genomes.jbrowse.org/demos/ for example

<img width="1861" height="710" alt="image" src="https://github.com/user-attachments/assets/9e76c2ae-3cb8-4ee8-bcd4-7cb28c3ad20f" />

Screenshot of hg38 vs hs1, as seen in the demos page

## Synteny and dotplot viewer improvements

We added some new features to the dotplot and synteny view including:

- Opacity/transparency slider
- Min length slider
- Easy picker for new color schemes
- Auto-diagonalization to reorder chromosomes
- "Location markers" for helping see where you are large alignments

<img width="1474" height="579" alt="image" src="https://github.com/user-attachments/assets/5e0148dc-15a9-4eb6-b587-6b26e3a7a1bb" />

<img width="1473" height="822" alt="image" src="https://github.com/user-attachments/assets/7bda3a16-9a4a-4e4c-af74-676a98aef301" />

A screenshot of the grape vs peach with "Color by Query" enabled in the dotplot
and synteny view

## Draw amino acids on Gene glyphs

The setting "Color by CDS" added the ability to draw the amino acid letters as
well (and was thus renamed "Color by CDS and draw amino acids"). This can help
with detailed investigations of protein structures, which is an area we are
actively developing (see https://github.com/GMOD/proteinbrowser for more
information)

<img width="1475" height="516" alt="image" src="https://github.com/user-attachments/assets/708a96b7-32a7-44f6-b095-7cd554af4a11" />

## Create 'Linked read display' for Alignments tracks

Users can now see paired-end reads and supplementary alignments as linked
entities using the "Linked reads display".

It has a variety of specialized features to help users dig into their data like

- Show/hide proper pairs
- Show/hide 'singletons' (reads without pairs or supplementary alignments)
- Coloring schemes that highlight mis-oriented pairs and abnormally large insert
  sizes
- And more

Example showing the "Linked reads display" with PacBio, Nanopore, and Illumina
paired-end reads. The long reads with split alignments are colored salmon and
lightblue so that you can see the flip in orientation from positive, to
negative, and back to positive again. The short paired-end reads are colored
navy and green to show the misoriented pairs

<img width="1460" height="961" alt="image" src="https://github.com/user-attachments/assets/48823471-0314-431b-8785-4e7100ff801e" />

## Improved modified reads plotting

Some of the code for handling modifications was given a fresh refactor, which
fixed bugs with 'duplex' data where modifications are stored on both a positive
and negative pass on the same read

Here is an example of of a screenshot showing chromatin accessibility via
introduced 6mA methylation from a Nanopore dataset (source:
https://epi2me.nanoporetech.com/chromatin-acc-hg002/)

<img width="1475" height="673" alt="image" src="https://github.com/user-attachments/assets/a627691b-a681-4302-9900-b99e772b8755" />

## Conclusion

Let us know if you have any feedback on these new features! Enjoy
