This is a big release on a number of fronts!

## Synteny and dotplot viewer improvements

We added a couple of new options for the dotplot and synteny view that have a
really positive impact on the rendered results including:

- Opacity/transparency slider
- Min length slider
- Easy picker for new color schemes
- Auto-diagonalization to reorder chromosomes

<img width="1474" height="579" alt="image" src="https://github.com/user-attachments/assets/5e0148dc-15a9-4eb6-b587-6b26e3a7a1bb" />

<img width="1473" height="822" alt="image" src="https://github.com/user-attachments/assets/7bda3a16-9a4a-4e4c-af74-676a98aef301" />

A screenshot of the grape vs peach with "Color by Query" enabled in the dotplot
and synteny view

## JBrowse 2 genome hubs

We created a new website called https://genomes.jbrowse.org that has JBrowse
instances for a huge number of species

This data is enabled by great work from UCSC to bulk load data from NCBI, which
we in turn imported to create a bunch of genome browsers

These browsers are available on the web and on the desktop app

Users can even browse "synteny views" between the different hubs by opening a
liftOver track, see https://genomes.jbrowse.org/demos/ for example

## Draw amino acids on Gene glyphs

The setting "Color by CDS" added the ability to draw the amino acid letters as
well (and was thus renamed "Color by CDS and draw amino acids"). This can help
with detailed investigations of protein structures, which is an area we are
actively developing (see https://github.com/GMOD/proteinbrowser for more
information)

<img width="1475" height="516" alt="image" src="https://github.com/user-attachments/assets/708a96b7-32a7-44f6-b095-7cd554af4a11" />

## Create 'Linked read display' for Alignments tracks

Users can now see paired-end reads and supplementary alignments as linked
entities using the "Linked reads display".

It has a variety of specialized features to help users dig into their data like

- Show/hide proper pairs
- Show/hide 'singletons' (reads without pairs or supplementary alignments)
- Coloring schemes that highlight mis-oriented pairs and abnormally large insert
  sizes
- And more

Example showing the Linked reads display with long and short read data. The long
reads with split alignments are colored salmon and lightblue so that you can see
the flip in orientation from positive, to negative, and then to positive again.
The short paired-end reads are colored navy and green to show the misoriented
pairs

<img width="1460" height="961" alt="image" src="https://github.com/user-attachments/assets/48823471-0314-431b-8785-4e7100ff801e" />
